himportscan/src/HImportScan/ImportScanner.hs

```diff
@@ -67,26 +65,55 @@ instance Aeson.ToJSON ModuleImport where
 scanImportsFromFile :: FilePath -> IO (Maybe ScannedImports)
 scanImportsFromFile filePath = do
   fileExists <- doesFileExist filePath
-  if fileExists
-  then Just . scanImports filePath <$> Text.readFile filePath
-  else pure Nothing
-
-scanImports :: FilePath -> Text -> ScannedImports
-scanImports filePath contents =
-  let preprocessedContents = Text.encodeUtf8 $ preprocessContents contents
-      sbuffer = case preprocessedContents of
-        PS ptr offset len -> StringBuffer ptr len offset
-      loc = mkRealSrcLoc (mkFastString filePath) 1 1
-   in case scanTokenStream filePath $ lexTokenStream sbuffer loc of
-    Left err -> error err
-    Right ScannedData{moduleName, importedModules, usesTH} ->
-      ScannedImports
-        { filePath = Text.pack filePath
-        , moduleName
-        , importedModules
-        , usesTH
-        }
-
+  if not fileExists
+  then pure Nothing
+  else scanImports filePath =<< Text.readFile filePath
+
+-- TODO[GL]: This function is only in IO because
+-- * we use printBagOfErrors to report an error, but we can easily factor that out
+-- * getImports is in IO, which in turn is only in IO to throw an error
+--   Perhaps we could raise an issue at ghc to make a pure variant.
+scanImports :: FilePath -> Text -> IO (Maybe ScannedImports)
+scanImports filePath contents = do
+  -- TODO[GL]: going through String just because StringBuffer doesn't have a Text interface is not the best
+  -- we could potentially skip that with more effort (e.g. go through ByteString, which is very similar to a StringBuffer)
+  let sb = GHC.stringToStringBuffer $ Text.unpack $ preprocessContents contents
+
+  -- TODO[GL]: Once we're on ghc 9.2 we can get rid of all the things relating to dynFlags, and use the much smaller
+  -- ParserOpts, as getImports no longer depends on DynFlags then.
+  let dynFlags = GHC.Utils.dynFlags
+
+  let
+    -- [GL] The fact that the resulting strings here contain the "-X"s makes me a bit doubtful that this is the right approach,
+    -- but this is what I found for now.
+    usesTH =
+      any (`elem` ["-XTemplateHaskell", "-XQuasiQuotes"]) $
+        map GHC.unLoc $
+          GHC.getOptions dynFlags sb filePath
+  -- TODO: should both of the files here be filePath?
```

Comment:
Not sure what the difference between the two arguments is here.

The docs on it say
```
           -> FilePath     -- ^ Filename the buffer came from.  Used for
                           --   reporting parse error locations.
           -> FilePath     -- ^ The original source filename (used for locations
                           --   in the function result)
```

Reply:
I think the second value shows up in the `Located` part of the result, while the former is used in error messages.

It could make a difference when using preprocessors (?). I don't have a strong grasp of this.